In [2]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import StandardScaler
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import roc_auc_score, confusion_matrix, classification_report
from sklearn.svm import LinearSVC
from sklearn import svm
from sklearn.model_selection import GridSearchCV, RandomizedSearchCV

In [2]:
#a)
train = pd.read_csv(r'C:\Users\u353822\Documents\GitHub\University-of-Chicago1\University of Chicago\Machine Learning\HW6\train_data.csv')
test = pd.read_csv(r'C:\Users\u353822\Documents\GitHub\University-of-Chicago1\University of Chicago\Machine Learning\HW6\test_data.csv')
print('Train shape: {}'.format(train.shape))
print('Test shape: {}'.format(test.shape))

Train shape: (507, 148)
Test shape: (168, 148)


In [29]:
#b)
train = train.dropna(how='any')
test = test.dropna(how='any')
train.head()

class  BrdIndx  Area  Round  Bright  Compact  ShpIndx  Mean_G  Mean_R  \
0  concrete      1.32   131   0.81  222.74     1.66     2.18  192.94  235.11   
1    shadow      1.59   864   0.94   47.56     1.41     1.87   36.82   48.78   
2    shadow      1.41   409   1.00   51.38     1.37     1.53   41.72   51.96   
3      tree      2.58   187   1.91   70.08     3.41     3.11   93.13   55.20   
4   asphalt      2.60   116   2.05   89.57     3.06     3.02   73.17   94.89   

   Mean_NIR    ...      SD_NIR_140  LW_140  GLCM1_140  Rect_140  GLCM2_140  \
0    240.15    ...           31.15    5.04       0.80      0.58       8.56   
1     57.09    ...           12.01    3.70       0.52      0.96       7.01   
2     60.48    ...           18.75    3.09       0.90      0.63       8.32   
3     61.92    ...           27.67    6.33       0.89      0.70       8.56   
4    100.64    ...           32.05    1.01       0.83      0.75       8.62   

   Dens_140  Assym_140  NDVI_140  BordLngth_140  GLCM3_140  
0      0.82       0.98     -0.10           1512    1287.52  
1      1.69       0.86     -0.14            196    2659.74  
2      1.38       0.84      0.10           1198     720.38  
3      1.10       0.96      0.20            524     891.36  
4      2.08       0.08     -0.10            496    1194.76  

[5 rows x 148 columns]

In [4]:
#c) 
X_train = train.drop('class',axis=1)
y_train = train['class']
X_test = test.drop('class',axis=1)
y_test = test['class']

In [5]:
#d)
sc = StandardScaler()
X_train_sc = sc.fit_transform(X_train)
X_test_sc = sc.transform(X_test)

In [6]:
#Part 2)
#a)
RFC = RandomForestClassifier()
RFC.fit(X_train_sc,y_train)

RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=None, max_features='auto', max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators=10, n_jobs=1,
            oob_score=False, random_state=None, verbose=0,
            warm_start=False)

In [7]:
#b & c)
y_pred = RFC.predict(X_test_sc)
print(confusion_matrix(y_test, y_pred))
print(classification_report(y_test, y_pred))

[[14  0  0  0  0  0  0  0  0]
 [ 1 21  0  3  0  0  0  0  0]
 [ 0  1 14  0  0  0  0  0  0]
 [ 0  4  0 19  0  0  0  0  0]
 [ 0  0  0  0 28  0  0  0  1]
 [ 1  0  1  0  0 13  0  0  0]
 [ 1  0  0  0  0  0 15  0  0]
 [ 0  1  0  6  2  0  0  5  0]
 [ 0  0  0  1  1  0  0  0 15]]
             precision    recall  f1-score   support

   asphalt        0.82      1.00      0.90        14
  building        0.78      0.84      0.81        25
       car        0.93      0.93      0.93        15
  concrete        0.66      0.83      0.73        23
     grass        0.90      0.97      0.93        29
      pool        1.00      0.87      0.93        15
    shadow        1.00      0.94      0.97        16
      soil        1.00      0.36      0.53        14
      tree        0.94      0.88      0.91        17

avg / total       0.88      0.86      0.85       168



In [8]:
#d)
y_pred_train = RFC.predict(X_train_sc)
print(confusion_matrix(y_train, y_pred_train))
print(classification_report(y_train, y_pred_train))

[[45  0  0  0  0  0  0  0  0]
 [ 0 96  0  1  0  0  0  0  0]
 [ 0  0 21  0  0  0  0  0  0]
 [ 0  3  0 90  0  0  0  0  0]
 [ 0  0  0  0 82  0  0  0  1]
 [ 0  0  0  0  0 14  0  0  0]
 [ 0  0  0  0  0  0 45  0  0]
 [ 0  2  0  1  0  0  0 17  0]
 [ 0  0  0  0  2  0  0  0 87]]
             precision    recall  f1-score   support

   asphalt        1.00      1.00      1.00        45
  building        0.95      0.99      0.97        97
       car        1.00      1.00      1.00        21
  concrete        0.98      0.97      0.97        93
     grass        0.98      0.99      0.98        83
      pool        1.00      1.00      1.00        14
    shadow        1.00      1.00      1.00        45
      soil        1.00      0.85      0.92        20
      tree        0.99      0.98      0.98        89

avg / total       0.98      0.98      0.98       507



In [9]:
pd.DataFrame(RFC.feature_importances_,index = X_train.columns,
        columns=['importance']).sort_values('importance',ascending=False).iloc[:5,:]

importance
Bright_40    0.061383
NDVI_40      0.047206
NDVI         0.039610
Mean_R_40    0.037660
NDVI_60      0.037371

In [10]:
# part 3
#a)
LSVC = LinearSVC()
LSVC.fit(X_train_sc, y_train)

LinearSVC(C=1.0, class_weight=None, dual=True, fit_intercept=True,
     intercept_scaling=1, loss='squared_hinge', max_iter=1000,
     multi_class='ovr', penalty='l2', random_state=None, tol=0.0001,
     verbose=0)

In [11]:
#b)
y_pred = LSVC.predict(X_test_sc)

In [12]:
#c)
print(classification_report(y_test,y_pred))
print(confusion_matrix(y_test,y_pred))

             precision    recall  f1-score   support

   asphalt        0.76      0.93      0.84        14
  building        0.65      0.88      0.75        25
       car        0.86      0.80      0.83        15
  concrete        0.79      0.65      0.71        23
     grass        0.72      0.90      0.80        29
      pool        1.00      0.87      0.93        15
    shadow        0.93      0.88      0.90        16
      soil        1.00      0.43      0.60        14
      tree        0.71      0.59      0.65        17

avg / total       0.80      0.78      0.77       168

[[13  0  0  0  0  0  1  0  0]
 [ 0 22  1  1  1  0  0  0  0]
 [ 0  2 12  0  0  0  0  0  1]
 [ 1  6  0 15  0  0  0  0  1]
 [ 0  0  0  1 26  0  0  0  2]
 [ 1  0  1  0  0 13  0  0  0]
 [ 2  0  0  0  0  0 14  0  0]
 [ 0  4  0  1  3  0  0  6  0]
 [ 0  0  0  1  6  0  0  0 10]]


In [13]:
#d)
y_pred_train = LSVC.predict(X_train_sc)
print(classification_report(y_train, y_pred_train))
print(confusion_matrix(y_train, y_pred_train))

             precision    recall  f1-score   support

   asphalt        1.00      1.00      1.00        45
  building        0.99      1.00      0.99        97
       car        1.00      1.00      1.00        21
  concrete        1.00      1.00      1.00        93
     grass        1.00      0.96      0.98        83
      pool        1.00      1.00      1.00        14
    shadow        1.00      1.00      1.00        45
      soil        1.00      1.00      1.00        20
      tree        0.98      1.00      0.99        89

avg / total       0.99      0.99      0.99       507

[[45  0  0  0  0  0  0  0  0]
 [ 0 97  0  0  0  0  0  0  0]
 [ 0  0 21  0  0  0  0  0  0]
 [ 0  0  0 93  0  0  0  0  0]
 [ 0  1  0  0 80  0  0  0  2]
 [ 0  0  0  0  0 14  0  0  0]
 [ 0  0  0  0  0  0 45  0  0]
 [ 0  0  0  0  0  0  0 20  0]
 [ 0  0  0  0  0  0  0  0 89]]


There does seem to be overfitting because our model returns an accuracy of 99% for our train data while the model return an accuracy of around 80% for our test data.

In [14]:
# part 4
#a)
LSVC = LinearSVC()
param_grid = {'C':np.arange(.01,10.2,.2)}

SVC = GridSearchCV(LSVC, param_grid=param_grid, cv=5)
SVC.fit(X_train_sc, y_train)

GridSearchCV(cv=5, error_score='raise',
       estimator=LinearSVC(C=1.0, class_weight=None, dual=True, fit_intercept=True,
     intercept_scaling=1, loss='squared_hinge', max_iter=1000,
     multi_class='ovr', penalty='l2', random_state=None, tol=0.0001,
     verbose=0),
       fit_params={}, iid=True, n_jobs=1,
       param_grid={'C': array([1.000e-02, 2.100e-01, 4.100e-01, 6.100e-01, 8.100e-01, 1.010e+00,
       1.210e+00, 1.410e+00, 1.610e+00, 1.810e+00, 2.010e+00, 2.210e+00,
       2.410e+00, 2.610e+00, 2.810e+00, 3.010e+00, 3.210e+00, 3.410e+00,
       3.610e+00, 3.810e+00, 4.010e+00, 4.210e+00, 4.410e+00, 4....00, 8.610e+00, 8.810e+00, 9.010e+00, 9.210e+00, 9.410e+00,
       9.610e+00, 9.810e+00, 1.001e+01])},
       pre_dispatch='2*n_jobs', refit=True, scoring=None, verbose=0)

In [15]:
#b)
print(SVC.best_params_)
print(SVC.best_estimator_)

{'C': 0.01}
LinearSVC(C=0.01, class_weight=None, dual=True, fit_intercept=True,
     intercept_scaling=1, loss='squared_hinge', max_iter=1000,
     multi_class='ovr', penalty='l2', random_state=None, tol=0.0001,
     verbose=0)


In [16]:
#c)
y_pred = SVC.best_estimator_.predict(X_test_sc)

In [17]:
#d)
print(classification_report(y_test,y_pred))
print(confusion_matrix(y_test,y_pred))

             precision    recall  f1-score   support

   asphalt        1.00      0.93      0.96        14
  building        0.63      0.88      0.73        25
       car        0.93      0.93      0.93        15
  concrete        0.62      0.70      0.65        23
     grass        0.73      0.83      0.77        29
      pool        1.00      0.87      0.93        15
    shadow        0.94      1.00      0.97        16
      soil        1.00      0.07      0.13        14
      tree        0.80      0.71      0.75        17

avg / total       0.81      0.78      0.76       168

[[13  0  0  0  0  0  1  0  0]
 [ 0 22  0  2  1  0  0  0  0]
 [ 0  1 14  0  0  0  0  0  0]
 [ 0  6  0 16  1  0  0  0  0]
 [ 0  0  0  2 24  0  0  0  3]
 [ 0  1  1  0  0 13  0  0  0]
 [ 0  0  0  0  0  0 16  0  0]
 [ 0  5  0  5  3  0  0  1  0]
 [ 0  0  0  1  4  0  0  0 12]]


In [18]:
#d)
y_pred_train = SVC.predict(X_train_sc)
print(classification_report(y_train, y_pred_train))
print(confusion_matrix(y_train, y_pred_train))

             precision    recall  f1-score   support

   asphalt        0.95      0.93      0.94        45
  building        0.81      0.91      0.86        97
       car        1.00      0.90      0.95        21
  concrete        0.87      0.86      0.86        93
     grass        0.84      0.82      0.83        83
      pool        1.00      0.86      0.92        14
    shadow        0.88      0.98      0.93        45
      soil        1.00      0.30      0.46        20
      tree        0.85      0.91      0.88        89

avg / total       0.87      0.87      0.86       507

[[42  0  0  0  0  0  3  0  0]
 [ 1 88  0  7  0  0  1  0  0]
 [ 0  1 19  0  0  0  0  0  1]
 [ 0 12  0 80  1  0  0  0  0]
 [ 0  1  0  2 68  0  0  0 12]
 [ 0  1  0  0  1 12  0  0  0]
 [ 1  0  0  0  0  0 44  0  0]
 [ 0  5  0  3  5  0  0  6  1]
 [ 0  0  0  0  6  0  2  0 81]]


There does seem to be overfitting because our model returns an accuracy of 87% for our train data while the model return an accuracy of around 80% for our test data.

In [19]:
#Part 5)
#a)
param_grid = {'C':np.arange(.01,10.2,.2),
             'degree': [2,3,4,5,6]}

SVC = GridSearchCV(svm.SVC(kernel='poly'), param_grid=param_grid, cv=5)
SVC.fit(X_train_sc, y_train)

GridSearchCV(cv=5, error_score='raise',
       estimator=SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape='ovr', degree=3, gamma='auto', kernel='poly',
  max_iter=-1, probability=False, random_state=None, shrinking=True,
  tol=0.001, verbose=False),
       fit_params={}, iid=True, n_jobs=1,
       param_grid={'C': array([1.000e-02, 2.100e-01, 4.100e-01, 6.100e-01, 8.100e-01, 1.010e+00,
       1.210e+00, 1.410e+00, 1.610e+00, 1.810e+00, 2.010e+00, 2.210e+00,
       2.410e+00, 2.610e+00, 2.810e+00, 3.010e+00, 3.210e+00, 3.410e+00,
       3.610e+00, 3.810e+00, 4.010e+00, 4.210e+00, 4.410e+00, 4.....010e+00, 9.210e+00, 9.410e+00,
       9.610e+00, 9.810e+00, 1.001e+01]), 'degree': [2, 3, 4, 5, 6]},
       pre_dispatch='2*n_jobs', refit=True, scoring=None, verbose=0)

In [20]:
#b)
print(SVC.best_params_)
print(SVC.best_estimator_)

{'C': 8.41, 'degree': 3}
SVC(C=8.41, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape='ovr', degree=3, gamma='auto', kernel='poly',
  max_iter=-1, probability=False, random_state=None, shrinking=True,
  tol=0.001, verbose=False)


In [21]:
#c)
y_pred = SVC.best_estimator_.predict(X_test_sc)

In [22]:
#d)
print(classification_report(y_test,y_pred))
print(confusion_matrix(y_test,y_pred))

             precision    recall  f1-score   support

   asphalt        0.87      0.93      0.90        14
  building        0.69      0.88      0.77        25
       car        1.00      0.73      0.85        15
  concrete        0.71      0.74      0.72        23
     grass        0.72      0.90      0.80        29
      pool        0.93      0.93      0.93        15
    shadow        0.88      0.88      0.88        16
      soil        0.25      0.07      0.11        14
      tree        0.87      0.76      0.81        17

avg / total       0.76      0.78      0.76       168

[[13  0  0  0  0  0  1  0  0]
 [ 0 22  0  2  1  0  0  0  0]
 [ 0  2 11  0  0  1  0  1  0]
 [ 0  5  0 17  0  0  0  1  0]
 [ 0  0  0  0 26  0  0  1  2]
 [ 0  0  0  0  0 14  1  0  0]
 [ 2  0  0  0  0  0 14  0  0]
 [ 0  3  0  4  6  0  0  1  0]
 [ 0  0  0  1  3  0  0  0 13]]


In [23]:
#e)
y_pred_train = SVC.predict(X_train_sc)
print(classification_report(y_train, y_pred_train))
print(confusion_matrix(y_train, y_pred_train))

             precision    recall  f1-score   support

   asphalt        1.00      1.00      1.00        45
  building        0.98      1.00      0.99        97
       car        1.00      0.95      0.98        21
  concrete        1.00      0.98      0.99        93
     grass        0.90      0.99      0.94        83
      pool        1.00      0.93      0.96        14
    shadow        1.00      1.00      1.00        45
      soil        1.00      0.75      0.86        20
      tree        1.00      0.99      0.99        89

avg / total       0.98      0.98      0.98       507

[[45  0  0  0  0  0  0  0  0]
 [ 0 97  0  0  0  0  0  0  0]
 [ 0  0 20  0  1  0  0  0  0]
 [ 0  1  0 91  1  0  0  0  0]
 [ 0  1  0  0 82  0  0  0  0]
 [ 0  0  0  0  1 13  0  0  0]
 [ 0  0  0  0  0  0 45  0  0]
 [ 0  0  0  0  5  0  0 15  0]
 [ 0  0  0  0  1  0  0  0 88]]


There does seem to be overfitting because our model returns an accuracy of 98% for our train data while the model return an accuracy of around 78% for our test data.


In [24]:
#Part 6)
#a)
param_grid = {'C':np.arange(.01,10.2,.2),
             'degree': [2,3,4,5,6]}

SVC = GridSearchCV(svm.SVC(kernel='rbf'), param_grid=param_grid, cv=5)
SVC.fit(X_train_sc, y_train)

GridSearchCV(cv=5, error_score='raise',
       estimator=SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape='ovr', degree=3, gamma='auto', kernel='rbf',
  max_iter=-1, probability=False, random_state=None, shrinking=True,
  tol=0.001, verbose=False),
       fit_params={}, iid=True, n_jobs=1,
       param_grid={'C': array([1.000e-02, 2.100e-01, 4.100e-01, 6.100e-01, 8.100e-01, 1.010e+00,
       1.210e+00, 1.410e+00, 1.610e+00, 1.810e+00, 2.010e+00, 2.210e+00,
       2.410e+00, 2.610e+00, 2.810e+00, 3.010e+00, 3.210e+00, 3.410e+00,
       3.610e+00, 3.810e+00, 4.010e+00, 4.210e+00, 4.410e+00, 4.....010e+00, 9.210e+00, 9.410e+00,
       9.610e+00, 9.810e+00, 1.001e+01]), 'degree': [2, 3, 4, 5, 6]},
       pre_dispatch='2*n_jobs', refit=True, scoring=None, verbose=0)

In [25]:
#b)
print(SVC.best_params_)
print(SVC.best_estimator_)

{'C': 5.21, 'degree': 2}
SVC(C=5.21, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape='ovr', degree=2, gamma='auto', kernel='rbf',
  max_iter=-1, probability=False, random_state=None, shrinking=True,
  tol=0.001, verbose=False)


In [26]:
#c)
y_pred = SVC.best_estimator_.predict(X_test_sc)

In [27]:
#d)
print(classification_report(y_test,y_pred))
print(confusion_matrix(y_test,y_pred))

             precision    recall  f1-score   support

   asphalt        0.93      0.93      0.93        14
  building        0.74      0.80      0.77        25
       car        0.94      1.00      0.97        15
  concrete        0.70      0.83      0.76        23
     grass        0.83      0.86      0.85        29
      pool        1.00      0.93      0.97        15
    shadow        0.88      0.94      0.91        16
      soil        0.80      0.29      0.42        14
      tree        0.83      0.88      0.86        17

avg / total       0.84      0.83      0.82       168

[[13  0  0  0  0  0  1  0  0]
 [ 0 20  1  3  1  0  0  0  0]
 [ 0  0 15  0  0  0  0  0  0]
 [ 0  4  0 19  0  0  0  0  0]
 [ 0  0  0  0 25  0  0  1  3]
 [ 0  0  0  0  0 14  1  0  0]
 [ 1  0  0  0  0  0 15  0  0]
 [ 0  3  0  4  3  0  0  4  0]
 [ 0  0  0  1  1  0  0  0 15]]


In [28]:
#e)
y_pred_train = SVC.predict(X_train_sc)
print(classification_report(y_train, y_pred_train))
print(confusion_matrix(y_train, y_pred_train))

             precision    recall  f1-score   support

   asphalt        1.00      1.00      1.00        45
  building        0.98      1.00      0.99        97
       car        1.00      1.00      1.00        21
  concrete        1.00      0.99      0.99        93
     grass        0.99      0.98      0.98        83
      pool        1.00      1.00      1.00        14
    shadow        1.00      1.00      1.00        45
      soil        1.00      1.00      1.00        20
      tree        0.99      0.99      0.99        89

avg / total       0.99      0.99      0.99       507

[[45  0  0  0  0  0  0  0  0]
 [ 0 97  0  0  0  0  0  0  0]
 [ 0  0 21  0  0  0  0  0  0]
 [ 0  1  0 92  0  0  0  0  0]
 [ 0  1  0  0 81  0  0  0  1]
 [ 0  0  0  0  0 14  0  0  0]
 [ 0  0  0  0  0  0 45  0  0]
 [ 0  0  0  0  0  0  0 20  0]
 [ 0  0  0  0  1  0  0  0 88]]


There does seem to be overfitting because our model returns an accuracy of 99% for our train data while the model return an accuracy of around 84% for our test data.


7) Conceptual questions
a) From the models run in steps 2-6, which performs the best based on the Classification Report? Support your reasoning with evidence around your test data
The last model utilizing a SVC with a kernel of rbf created the best model when utilizing the classificaiton reports. We are able to make this claim due to the fact that the model returned the highest accuracy for our test data (88%) and our training data (99%).

b) Compare models run for steps 4-6 where different kernels were used. What is the benefit of using a polynomial or rbf kernel over a linear kernel? What could be a downside of using a polynomial or rbf kernel? 
Utilizing a polynomial kernel or a rbf kernel allows us to create non linear hyperplanes which are otherwise impossible in a linear SVC. A potential downside to using a polynomial or rbf kernel is that those models typically take longer to run when compared to a linaer SVC model.

c) Explain the 'C' parameter used in steps 4-6. What does a small C mean versus a large C in sklearn? Why is it important to use the 'C' parameter when fitting a model? 
The C parameter tells the SVM how much you want the model to avoid misclassification. If a larger C is inputted into the model, the model will look for a smaller-margin hyperplane if that hyperplane does a better job correctly classifying the data. For a small value of c, the optomizer will look for a larger margin seperating the hyperplane therfore allowing for misclassification. The C parameter is very important for SVC because you are telling the optimizer how leanient the model can be with misclassifications. 

d)Scaling our input data does not matter much for Random Forest, but it is a critical step for Support Vector Machines. Explain why this is such a critical step. Also, provide an example of a feature from this data set that could cause issues with our SVMs if not scaled.
SVM optimizes based on the distribution of points. If there are parameters in the model with very different scales, and the data wasn't scaled, the SVM wouldn't be able to take into account the true distributions and variance of the data because it would be thrown off by the scale of the parameters. Before the data is scaled, both BrdIndx and Area have very different scales. When we scale these parameters to have a mean of 0 and variance of 1, it allows SVM to understand the true distribution of data between these two parameters. 

e) Describe conceptually what the purpose of a kernel is for Support Vector Machines.
The purpose of a kernel is to create another dimension ontop of the dimensionality the data provides which allows for the model to create non linear hyperplanes onto the linear space of the data.